flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/stream/PushCalcPastChangelogNormalizeRuleTest.java

```diff
@@ -37,43 +38,155 @@
 
-    @BeforeEach
-    void before() {
-        util = streamTestUtil(TableConfig.getDefault());
-    }
-
-    @Test
-    void testWithSinglePrimaryKeyFilter() {
-        final TableDescriptor sourceDescriptor =
+    @BeforeAll
+    static void setup() {
+        sourceDescriptor =
                 TableFactoryHarness.newBuilder()
                         .schema(
                                 Schema.newBuilder()
-                                        .column("f0", STRING())
+                                        .column("f0", INT())
                                         .column("f1", INT().notNull())
+                                        .column("f2", STRING())
+                                        .column("f3", BIGINT().notNull())
                                         .primaryKey("f1")
                                         .build())
                         .unboundedScanSource(ChangelogMode.upsert())
                         .build();
 
-        util.tableEnv().createTable("T", sourceDescriptor);
-        util.verifyRelPlan("SELECT * FROM T WHERE f1 < 1");
-    }
-
-    @Test
-    void testWithMultipleFilters() {
-        final TableDescriptor sourceDescriptor =
+        sourceDescriptorWithTwoPrimaryKeys =
                 TableFactoryHarness.newBuilder()
                         .schema(
                                 Schema.newBuilder()
                                         .column("f0", STRING())
                                         .column("f1", INT().notNull())
-                                        .column("f2", STRING())
-                                        .primaryKey("f1")
+                                        .column("f2", BIGINT().notNull())
+                                        .column("f3", STRING())
+                                        .column("f4", BIGINT().notNull())
+                                        .column("f5", BIGINT().notNull())
+                                        .column("f6", BIGINT().notNull())
+                                        .column("f7", BIGINT().notNull())
+                                        .primaryKey("f1", "f2")
                                         .build())
                         .unboundedScanSource(ChangelogMode.upsert())
                         .build();
+    }
+
+    @BeforeEach
+    void before() {
+        util = streamTestUtil(TableConfig.getDefault());
+    }
+
+    @Test
+    void testWithSinglePrimaryKeyFilter() {
+        util.tableEnv().createTable("T", sourceDescriptor);
+        util.verifyRelPlan("SELECT * FROM T WHERE f1 < 1");
+    }
+
+    @Test
+    void testPrimaryKeySeveralSameSources() {
+        util.tableEnv().createTable("T", sourceDescriptor);
+        // Shouldn't be pushed down since there is no common filter for ChangelogNormalize with the
+        // same source
+        // verifyExecPlan is intended here as it will show whether the node is reused or not
+        util.verifyExecPlan(
```

Comment:
Am I missing something? I can see these tests would drive various cases of re-use or pushdown, but I am not seeing the part of the test that proves the appropriate reuse or pushdown has actually occurred.

Reply:
do you see `+- Reused(reference_id=[1])` in  plans?
This indicates that reuse happens

Reply:
depending on conditions filter is pushdown "below" changelognormalize or stayed as is or splitted and partially pushdown "below" changelognormalize